chemprop/models/ffn.py

```diff
@@ -0,0 +1,439 @@
+        """
+        a_hidden, a_scope, b_hidden, b_scope, b2br = input
+
+        if self.ffn_type == "atom":
+            hidden = a_hidden
+            scope = a_scope
+        elif self.ffn_type == "bond":
+            forward_bond = b_hidden[b2br[:, 0]]
+            backward_bond = b_hidden[b2br[:, 1]]
+            scope = [((start - 1) // 2, size // 2) for start, size in b_scope]
+        else:
+            raise RuntimeError(f"Unrecognized ffn_type of {self.ffn_type}.")
+
+        if self.ffn_type == "atom":
+            output = self.ffn_readout(hidden)
+            output = output[1:]  # remove the first one which is zero padding
+        elif self.ffn_type == "bond":
+            b_hidden_1 = torch.cat([forward_bond, backward_bond], dim=1)
+            b_hidden_2 = torch.cat([backward_bond, forward_bond], dim=1)
+            output_1 = self.ffn_readout(b_hidden_1)
+            output_2 = self.ffn_readout(b_hidden_2)
+            output = (output_1 + output_2) / 2
+            if bond_types is not None:
+                output = output + bond_types.reshape(-1, 1)
+
+        return output
+
+class FFNAtten(nn.Module):
+    """
+    A :class:`FFNAtten` is a multiple feed forward neural networks (FFN) to predict
+    the atom/bond descriptors with applying constraint on output. An attention-based
+    constraint is used. This method is from `Regio-selectivity prediction with a
+    machine-learned reaction representation and on-the-fly quantum mechanical descriptors
+    <https://pubs.rsc.org/en/content/articlelanding/2021/sc/d0sc04823b>`_, section 2.2.
+    """
+
+    def __init__(
+        self,
+        features_size: int,
+        hidden_size: int,
+        num_layers: int,
+        output_size: int,
+        dropout: float,
+        activation: nn.Module,
+        ffn_base: nn.Module,
+        ffn_type: str = "atom",
+        shared_ffn: bool = True,
+        weights_ffn_num_layers: int = 2,
+    ):
+        """
+        :param features_size: Dimensionality of input features.
+        :param hidden_size: Dimensionality of hidden layers.
+        :param num_layers: Number of layers in FFN.
+        :param output_size: The size of output.
+        :param dropout: Dropout probability.
+        :param activation: Activation function.
+        :param ffn_base: The shared base layers (all but the last) of the FFN between tasks.
+        :param ffn_type: The type of target (atom or bond).
+        :param shared_ffn: Whether to share weights in the ffn between different atom tasks and bond tasks.
+        :param weights_ffn_num_layers: Number of layers in FFN for determining weights used to correct the constrained targets.
+        """
+        super().__init__()
+
+        if num_layers == 1:
+            base_output_size = features_size
+        else:
+            base_output_size = hidden_size
+
+        if shared_ffn:
+            self.ffn = ffn_base
```

Comment:
can we condense these two a single argument of type `Optional[Module]`. I.e., **if** an `ffn_base` is passed in, this class will use it

Reply:
This is a good suggestion. I will apply it. Thank you!